modules/core/arrow-test/src/main/kotlin/arrow/test/laws/BracketLaws.kt

```diff
@@ -154,15 +137,64 @@ object BracketLaws {
         .equalUnderTheLaw(just(expected), EQ)
     }
 
-  fun <F> Bracket<F, Throwable>.guaranteeMustRunFinalizerOnError(
+  fun <F> Bracket<F, Throwable>.guaranteeCaseMustRunFinalizerOnError(
     EQ: Eq<Kind<F, Int>>
   ): Unit =
     forAll(Gen.int(), Gen.throwable()) { i, t ->
       val msg = AtomicIntW(0)
       raiseError<Int>(t)
-        .guarantee(unit().map { msg.value = i })
+        .guaranteeCase { unit().map { msg.value = i } }
         .attempt()
         .map { msg.value }
         .equalUnderTheLaw(just(i), EQ)
     }
+
+  fun <F> Bracket<F, Throwable>.bracketIsDerivedFromBracketCase(
+    EQ: Eq<Kind<F, Int>>
+  ): Unit =
+    forAll(Gen.int().applicativeError(this)) { fa: Kind<F, Int> ->
+      fa.bracket(release = { just<Unit>(Unit) }, use = { just(it) }).equalUnderTheLaw(fa.bracketCase(release = { _, _ -> just<Unit>(Unit) }, use = { just(it) }), EQ)
+    }
+
+  fun <F> Bracket<F, Throwable>.guaranteeIsDerivedFromBracket(
+    finalizer: Kind<F, Unit>,
+    EQ: Eq<Kind<F, Int>>
+  ): Unit =
+    forAll(Gen.int().applicativeError(this)) { fa: Kind<F, Int> ->
+      fa.guarantee(finalizer).equalUnderTheLaw(just(Unit).bracket({ finalizer }, use = { fa }), EQ)
+    }
+
+  fun <F> Bracket<F, Throwable>.guaranteeCaseIsDerivedFromBracketCase(
+    finalizer: (ExitCase<Throwable>) -> Kind<F, Unit>,
+    EQ: Eq<Kind<F, Int>>
+  ): Unit =
+    forAll(Gen.int().applicativeError(this)) { fa: Kind<F, Int> ->
+      fa.guaranteeCase(finalizer).equalUnderTheLaw(just(Unit).bracketCase({ _, e -> finalizer(e) }) { fa }, EQ)
+    }
+
+  fun <F> Bracket<F, Throwable>.onCancelIsDerivedFromGuaranteeCase(
+    GENK: GenK<F>,
+    EQ: Eq<Kind<F, Int>>
+  ): Unit =
+    forAll(Gen.int().applicative(this), GENK.genK(Gen.unit())) { fa: Kind<F, Int>, ff: Kind<F, Unit> ->
+      fa.onCancel(ff).equalUnderTheLaw(fa.guaranteeCase {
+        when (it) {
+          is ExitCase.Canceled -> ff
+          else -> unit()
+        }
+      }, EQ)
+    }
+
+  fun <F> Bracket<F, Throwable>.onErrorIsDerivedFromGuaranteeCase(
+    GENK: GenK<F>,
+    EQ: Eq<Kind<F, Int>>
+  ): Unit =
+    forAll(Gen.throwable().raiseError<F, Int, Throwable>(this), Gen.functionAToB<Throwable, Kind<F, Unit>>(GENK.genK(Gen.unit()))) { fe: Kind<F, Int>, ff: (Throwable) -> Kind<F, Unit> ->
```

Comment:
You're always comparing the result of `{ _ -> unit() }` with each-other.
I think here you'd want a test similar to `guaranteeCaseMustRunFinalizerOnError` but for `onError` rather then checking if `onError` was dereived from `guaranteeCaseMustRunFinalizerOnError`

Reply:
not in this case, because `Gen.throwable().raiseError` will always emit an error, therefore the function is called. If i replace the `ff(it.e)` in the error case for something else the test will actually fail